### ISSUE TITLE
More principled WMS version negotation

### ISSUE THREAD

**Issue by prl900** [TerriaJS/terriajs#2590]
I'm not sure yet if this is a TerriaJS or a server side problem. But I've observed a non expected behaviour of TerriaJS interacting with our server.

TerriaJS interrogates our server with the following WMS GetCapabilities request:
http://gsky-test.nci.org.au/ows?service=WMS&request=GetCapabilities&version=1.3.0&tiled=true

As far as I can understand the server responds with a XML document stating that WMS is version 1.3.0

But then, TerriaJS sends requests using WMS 1.1.1 as this one:
http://gsky.nci.org.au/ows?time=2017-03-06T00%3A00%3A00.000Z&srs=EPSG%3A3857&transparent=true&format=image%2Fpng&exceptions=application%2Fvnd.ogc.se_xml&styles=&tiled=true&feature_count=101&service=WMS&version=1.1.1&request=GetMap&layers=global%3Atotalcover&bbox=0%2C0%2C5009377.085697312%2C5009377.085697312&width=256&height=256

I'd like to know if this change between versions is related to TerriaJS or a consequence of a problem in the returned GetCapabilities document.

**Comment by kring** [comment]
TerriaJS always requests version `1.1.1` for GetMap requests.  You can change this if necessary by adding `"version": "1.3.0"` to the `parameters` property in the init file.

The reason it does this is because Geoserver, which is far and away the most popular WMS server we talk to, only supports `tiled=true` for WMS 1.1.1 requests.  Setting `tiled=true` makes Geoserver go through the GeoWebCache, if one is configured, drastically improving performance.  Also, I'm not aware of any server that fails to answer WMS 1.1.1 requests.

**Comment by prl900** [comment]
Thanks @kring for looking into this. 

What I find confusing is that although TerriaJS starts interrogating a WMS with version 1.3.0 and the server responds with a 1.3.0 GetCapabilities document, it assumes that the server has 1.1.1 capabilities for GetMap requests. I'm not sure how GeoServer handles this, but I reckon it could be a problem for servers that only implement WMS 1.3.0 (ie SRS vs CRS / axis coordinates order)

As far as we are concerned, I understand that we just need to make sure our server exposes both 1.1.1 and 1.3.0 to assure compatibility with TerriaJS.

EDIT: Also, I've just noticed that the GetCapabilities requests that we receive combine version=1.3.0 and tiled=true:
http://gsky-test.nci.org.au/ows?service=WMS&request=GetCapabilities&version=1.3.0&tiled=true
As you say, if tiled=true is only available in 1.1.1, should this request return an error?

**Comment by kring** [comment]
Yeah, I agree it's pretty weird and seems kind of buggy.  We've gotten away with it because we've never tried to connect to a server that only supports 1.3.0.

As for GetCapabilities 1.3.0 and tiled=true, you're right that it's not supported, but most (all?) WMS servers ignore it rather than returning an error.  We should remove it, since it's not doing anything productive.  We used to request 1.1.1 for GetCapabilities as well.  We changed that after seeing some WMS servers fail to respond to a 1.1.1 GetCapabilities request.  These servers (ArcGIS specifically) would return an error, rather than returning the 1.3.0 document as the WMS spec says they should, and despite the fact that ArcGIS supports 1.1.1 for GetMap just fine.  I believe the problem was that ArcGIS requires the GetCapabilities document to be explicitly provided for each version, and the administrators never bothered to set up anything but 1.3.0.

TL;DR: supporting a wide range of WMS servers is a tricky balancing act.  TerriaJS's WMS support is known to work with a wide range of WMS servers out there, but admittedly it does some weird stuff along the way.

What would a more correct solution look like?  Perhaps:
1. Request GetCapabilities 1.1.1.
2. If we get back a 1.1.1 document with tiling information, use GetMap with version 1.1.1 and tiled=true.
3. If we get an error or if there's no tiling information, request GetCapabilities at 1.3.0.
4. If both GetCapabilities requests return an error, give up.
5. If 1.3.0 returns an error but 1.1.1 doesn't, use GetMap with version 1.1.1.
6. If the 1.3.0 request returns a 1.1.1 document, use GetMap with version 1.1.1.
7. Otherwise, use GetMap with version 1.3.0.

That's just off the top of my head and probably wrong, but hopefully not too far off.  Other than complexity, this would also involve an extra GetCapabilities request in common scenarios.

Anyway, I'd suggest supporting 1.1.1 GetMap requests for best compatibility in general, not just with TerriaJS.  I don't think there are significant differences in that service that should present much problem for you.
